### PR TITLE
fp16 support

### PR DESCRIPTION
**Pull request by lukasc-ch**
This PR includes fp16 support for the following modules: 
- SpatialConvolution
- all point-wise activations
- all 2d pooling operations
- all softmax
- SpatialCrossMapLRN
- SpatialDivisiveNormalization

Requires merge of torch/cutorch#394 first.

### CODE REVIEW

File: Pointwise.lua [hunk 1]
```diff
@@ -31,11 +31,14 @@ function Pointwise:createIODescriptors(input)
    end
 
    local nElem = input:nElement()
-   self.nElem = self.nElem or nElem -- this goes to the second branch only once
-   if self.iDesc and nElem == self.nElem then return end
-   self.nElem = nElem
-   self.iDesc = cudnn.toDescriptor(input:view(1,1,1,nElem))
-
+   self.nElem = self.nElem or nElem
+   if self.iDesc and nElem == self.nElem and
```
Comment (apaszke) [inline Pointwise.lua]: Input tensor can change between runs, so I think you can't omit creating a descriptor for it.

File: SpatialCrossMapLRN.lua [hunk 2]
```diff
@@ -33,9 +33,12 @@ function LRN:createIODescriptors(input)
    end
    assert(input:dim() == 4 and input:isContiguous());
    if not self.iDesc or
-      input:size(1) ~= self.iSize[1] or input:size(2) ~= self.iSize[2]
-   or input:size(3) ~= self.iSize[3] or input:size(4) ~= self.iSize[4] then
+      input:size(1) ~= self.iSize[1] or input:size(2) ~= self.iSize[2] or
+      input:size(3) ~= self.iSize[3] or input:size(4) ~= self.iSize[4] or 
+   not self.iType or input:type() ~= self.iType then
```
Comment (apaszke) [inline SpatialCrossMapLRN.lua]: Is this `not` necessary? If `self.iType == nil` the next condition will evaluate to `true` anyway

File: Pointwise.lua (hunk 1 above)
Comment (lukasc-ch) [inline Pointwise.lua]: If the dimensions of the input tensor are changed, or (new) the type is changed, a new descriptor is created. If both these properties are the same, the same descriptor can be reused. Except for the type part it should be exactly the same as before.

File: SpatialConvolution.lua [hunk 3]
```diff
@@ -361,7 +383,9 @@ local function makeContiguous(self, input, gradOutput)
 end
 
 function SpatialConvolution:updateOutput(input)
-    if not self.weightDesc then self:resetWeightDescriptors() end
+    if not self.weightDesc or not self.weightType or self:type() ~= self.weightType then
```
Comment (apaszke) [inline SpatialConvolution.lua]: `not` condition unnecessary

File: Pointwise.lua (hunk 1 above)
Comment (apaszke) [inline Pointwise.lua]: Doesn't a descriptor hold a pointer to Tensor's data? You don't even know if the tensor you have received as input previously haven't been freed in the meantime.

File: SpatialCrossMapLRN.lua (hunk 2 above)
Comment (lukasc-ch) [inline SpatialCrossMapLRN.lua]: You are right.

File: Pooling.lua [hunk 4]
```diff
@@ -50,33 +50,36 @@ function Pooling:createIODescriptors(input)
...
-         oW = math.ceil((input:size(4)+self.padW*2 - self.kW)/self.dW + 1)
-         oH = math.ceil((input:size(3)+self.padH*2 - self.kH)/self.dH + 1)
-      else
-         oW = math.floor((input:size(4)+self.padW*2 - self.kW)/self.dW + 1)
-         oH = math.floor((input:size(3)+self.padH*2 - self.kH)/self.dH + 1)
-      end
-      self.output:resize(input:size(1), input:size(2), oH, oW)
+      input:size(1) ~= self.iSize[1] or input:size(2) ~= self.iSize[2] or
+      input:size(3) ~= self.iSize[3] or input:size(4) ~= self.iSize[4] or 
+      not self.iType or input:type() ~= self.iType then
```
Comment (apaszke) [inline Pooling.lua]: `not` unnecessary

**Comment by apaszke** [comment]
Does cuDNN support operations on fp16 weights and float inputs?

File: SpatialConvolution.lua [hunk 5]
```diff
@@ -26,14 +26,31 @@ function SpatialConvolution:__init(nInputPlane, nOutputPlane,
     self:reset()
     -- should nil for serialization, the reset will still work
     self.reset = nil
+
+    self._halfPrecision = false
+    devProps = cutorch.getDeviceProperties(cutorch.getDevice())
+    -- half-precision computing (not i/o only) needs HW support (Tegra X1 or Pascal)
```
Comment (apaszke) [inline SpatialConvolution.lua]: Is this really true? As far as I remember current GPUs can do FP16 computation as well, it's just that they don't have any special HW support for it, so it won't be much faster.

**Comment by szagoruyko** [comment]
@apaszke no, everything should be the same type

File: SpatialConvolution.lua (hunk 5 above)
Comment (apaszke) [inline SpatialConvolution.lua]: It might be enough to check `cutorch.hasHalf`

File: Pointwise.lua (hunk 1 above)
Comment (lukasc-ch) [inline Pointwise.lua]: No, it doesn't hold a pointer to the data.... You can see it also by the signature of cudnnSetTensorNdDescriptor, as called within init.lua for cudnn.toDescriptor(t): 
```
   errcheck('cudnnSetTensorNdDescriptor', descriptor[0], cudnnDataType,
            t:dim(), size:data(), stride:data())
```

Comment (apaszke) [inline Pointwise.lua]: Ok, my bad then. Sorry

**Comment by szagoruyko** [comment]
for reference, here is my version that I rebased on R5 https://github.com/soumith/cudnn.torch/compare/R5...szagoruyko:fp16
I will do a PR when I finish writing tests for all modules.

**Comment by apaszke** [comment]
Hm, I think this implementation won't throw any errors if you have such situation. It will just reinitialise input descriptors without checking if their type matches the weight type.

**Comment by apaszke** [comment]
@szagoruyko +1 For `cudnn.typemap`

**Comment by lukasc-ch** [comment]
@apaszke You actually cannot compute with fp16 on pre-Pascal GPUs except for the X1. cuDNN supports the following configs: 
 - double data, double computations
 - single data, single computation
 - half data, single computation
 - half data, half computation

**Comment by apaszke** [comment]
@lukasc-ch Well, there are 3 types there: input, weight and computation. I agree that cuDNN doesn't support halfs for computation on pre-Pascal architectures, but my question was, what happens if you have float input and half weights - is this supported? It doesn't seem to be checked by your implementation.

**Comment by soumith** [comment]
The dilemma that I have now is that there are two PRs (which is why it is nice to pre-declare that you are working on the feature). One by @lukasc-ch and other by @szagoruyko . Sergey's is much cleaner in design, it introduces a typemap and shape that keep the rest of the code fairly generic. However it is not complete yet. I think we can integrate the needed parts of @lukasc-ch 's PR into Sergey's and merge it in.
Does everyone else on the thread agree?

**Comment by lukasc-ch** [comment]
@soumith yes, this makes sense :+1: 
@apaszke no, this is not supported, at least not yet :)

**Comment by lukasc-ch** [comment]
PR #206 replaces this. Closing it.

**Pull request closed without merging**